### PR TITLE
Fixed undetected unused stubbing when matching previous stubbed call

### LINKED ISSUE #1846: Broken unused stubbing reporting when matching previous stubbed call
When stubbing the same method call twice, and the second stub setup matches to first stubbed call, the first stub is not reported as unused, because Mockito detects the next stub setup as a usage of the previous stub setup. This is an example test:
```
public class UnnecessaryStubbingWithImplicitEqMatchers {
    @Mock IMethods mock;
    MockitoSession mockito = Mockito.mockitoSession().initMocks(this).strictness(Strictness.STRICT_STUBS).startMocking();

    @After public void after() {
        mockito.finishMocking();
    }

    @Test public void unnecessary_stubbing() {
        given(mock.simpleMethod("1")).willReturn("one");
        given(mock.simpleMethod("2")).willReturn("three");
        given(mock.simpleMethod("1")).willReturn("two");
        given(mock.simpleMethod("1")).willReturn("four");

        mock.simpleMethod("1");
        mock.simpleMethod("2");
    }
}
```

The test above should fail with `UnnecessaryStubbingException` containing two unused stubs, but it does not.

Mockito version: v3.2.4

### PR DESCRIPTION
**Pull request by andreisilviudragnea**
Fixes #1846

### CODE REVIEW

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java [hunk 1]
```diff
@@ -93,6 +99,25 @@ public void run() {
         }
     }
 
+    public static class UnnecessaryStubbingWithImplicitEqMatchers {
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: The name does not seem correct - we don't use explicit 'eq' matchers in those tests.

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java [hunk 2]
```diff
@@ -93,6 +99,25 @@ public void run() {
...
+        @Mock IMethods mock;
+        MockitoSession mockito = Mockito.mockitoSession().initMocks(this).strictness(Strictness.STRICT_STUBS).startMocking();
+
+        @After public void after() {
+            mockito.finishMocking();
+        }
+
+        @Test public void unnecessary_stubbing() {
+            given(mock.simpleMethod("1")).willReturn("one");
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: Those are neat tests! In addition, I'd love to see coverage for the following use case:

```
when(mock.foo(anyString()).thenReturn(x);
when(mock.foo("foo")).thenReturn(x);
```

In above, we also should report 2 unused stubs. This way, we cover the case with arg matcher.

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java [hunk 3]
```diff
@@ -42,6 +42,12 @@
                 .fails("unnecessary_stubbing", UnnecessaryStubbingException.class);
     }
 
+    @Test public void unnecessary_stubbing_with_eq_implicit_matchers() {
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: This is a good e2e test.

Can you add coverage for the same use case but with the "willReturn(x).when(...)" stubbing style? This should "just" work without new changes but adding extra coverage would be nice for our codebase :-)

File: src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java [hunk 4]
```diff
@@ -52,10 +53,35 @@ public void resetInvocationForPotentialStubbing(MatchableInvocation invocationMa
...
+    private void unmarkLastUsedStubIfNecessary() {
+        List<Invocation> all = registeredInvocations.getAll();
+        if (all.isEmpty()) {
+            return;
+        }
+
+        Invocation lastInvocation = all.get(all.size() - 1);
+
+        for (StubbedInvocationMatcher s : stubbed) {
+            if (Arrays.equals(s.getInvocation().getRawArguments(), lastInvocation.getRawArguments())) {
```
[the remark this answers is not in the document]
Comment (andreisilviudragnea) [reply src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java]: Done.

File: src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java [hunk 5]
```diff
@@ -52,10 +53,35 @@ public void resetInvocationForPotentialStubbing(MatchableInvocation invocationMa
...
+     *  usage of the first stubbed call. When the second stub is set up, the previous one
+     *  is marked as unused again.
+     */
+    private void unmarkLastUsedStubIfNecessary() {
+        List<Invocation> all = registeredInvocations.getAll();
+        if (all.isEmpty()) {
+            return;
+        }
+
+        Invocation lastInvocation = all.get(all.size() - 1);
```
[the remark this answers is not in the document]
Comment (andreisilviudragnea) [reply src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java]: Done.

File: src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java [hunk 6]
```diff
@@ -52,10 +53,35 @@ public void resetInvocationForPotentialStubbing(MatchableInvocation invocationMa
...
+
+        for (StubbedInvocationMatcher s : stubbed) {
+            if (Arrays.equals(s.getInvocation().getRawArguments(), lastInvocation.getRawArguments())) {
+                s.markStubUsed(null);
+                lastInvocation.markStubbed(null);
```
[the remark this answers is not in the document]
Comment (andreisilviudragnea) [reply src/main/java/org/mockito/internal/stubbing/InvocationContainerImpl.java]: I have updated the javadoc of the two methods.

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java (hunk 1 above)
Comment (andreisilviudragnea) [reply src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: Done.

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java (hunk 2 above)
Comment (andreisilviudragnea) [reply src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: If the test contains, for example, the `mock.foo("bar");` call, the test should report only one unused stubbing (the one with the `"foo"` argument). What did you mean by 2 unused stubs in this case?

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java (hunk 3 above)
Comment (andreisilviudragnea) [reply src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: Done.

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java (hunk 2 above)
Comment (mockitoguy) [reply src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: I meant this:

```
when(mock.foo(anyString()).thenReturn(x);
when(mock.foo("foo")).thenReturn(x);
//2 unused stubbings
```

```
when(mock.foo(anyString()).thenReturn(x);
when(mock.foo("foo")).thenReturn(x);

mock.foo("foo")
//1 unused stubbing
```

Hope it makes sense!

**Comment by mockitoguy** [comment]
I'll review by tomorrow, thank you for the updates!

File: src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java (hunk 2 above)
Comment (andreisilviudragnea) [reply src/test/java/org/mockitousage/stubbing/StrictStubbingEndToEndTest.java]: Got it.

**Comment by mockitoguy** [comment]
I tweaked the tests. When the build is green I'll merge.

Thank you for this nice PR!!!

**Pull request merged**